dynamic_instrumentation/src/instrument.rs

```diff
@@ -117,6 +121,239 @@ fn is_region_or_unsafe_ptr(ty: Ty) -> bool {
+                        args: _,
+                        destination,
+                        target,
+                        ..
+                    } => {
+                        if destination
+                            .as_local()
+                            .filter(|l| self.address_taken.contains(l))
+                            .is_some()
+                        {
+                            if let Some(next_block) = target {
+                                // put into first statement of following block
+                                local_address_statements.push((
+                                    *destination,
+                                    Location {
+                                        block: *next_block,
+                                        statement_index: 0,
+                                    },
+                                ));
+                            }
+                        }
+                    }
+                    TerminatorKind::DropAndReplace {
+                        place,
+                        value,
+                        target,
+                        unwind: _,
+                    } if value.place().is_some() => {
+                        if let Some(local_to_address) =
+                            place.as_local().and_then(|l| self.local_to_address.get(&l))
+                        {
+                            // put into first statement of following block
+                            local_address_statements.push((
+                                *place,
+                                Location {
+                                    block: *target,
+                                    statement_index: 0,
+                                },
+                            ));
+                            self.local_to_address.insert(place.local, *local_to_address);
+                        }
+                    }
+                    _ => (),
+                }
+            }
+        }
+
+        // END ADDRESS-TAKING STATEMENTS
+        // visit places first before inserting `_y = &raw _x`, so that we don't accidentally
+        // rewrite `_y = &raw _x` into `y = &raw (*_y)`
+        self.super_body(body);
+
+        // sort the insertion locations by statement id descending so that statement insertion
+        // for statements with indices N and M, where N < M, does not shift/invalidate index M
+        local_address_statements.sort_by_key(
+            |(
+                _,
+                Location {
+                    block: _bid,
+                    statement_index: sid,
+                },
+            )| *sid,
+        );
+        let local_address_statements = local_address_statements.into_iter().rev();
+
+        // Add `_y = &raw _x` for each address-taken local _x
+        for (
+            addr_taken_local,
+            Location {
+                block: bid,
+                statement_index: sid,
+            },
```

Comment:
I'd just leave this as `loc` and use `loc.block` and `loc.statement_index`, as those are also clearer to understand than `bid` and `sid`.